README.md
device-checkout-rs
==================
Reimplementation/Distant ancestor of https://github.com/tismith/deviceCheckout. 
Written in rust. The HTTP API endpoints could use some more 
breadth, but the form based web ui is functional.

Since this is using `rocket` for the web framework, we need to use rust nightly, so we've pinned a working compiler using the rustc-toolchain file. Cargo build will pull down and install the correct compiler.

We use `diesel-migrations` to automatically build and migrate the database. No need to seed the database manually.

We are using:
-------------
* `rocket` for the web framework
* `diesel` as the database abstraction and orm
* `serde` for json serialization/deserialization
* `log` and `stderrlog` for configurable logging macros
* `clap` for commandline argument processing
* `failure` for error handling
* `assert_cli` for integration testing

Installation:
=============

Using `cargo`:
--------------

Install rustup and cargo: https://www.rust-lang.org/tools/install

```sh
cargo build
cargo run
```

Access help via command line args: `cargo run -- -h`

Using `snap`:
-------------

We're using [snapcraft](https://build.snapcraft.io) to automatically build snaps of device-checkout.

```sh
sudo snap install device-checkout
```

Using `docker`:
---------------

```sh
#Runs device-checkout on port 1234 with the database at /var/lib/devices.db
docker run -p 1234:8000 -v /var/lib:/var/lib/device-checkout tismith/device-checkout-rs
```

Development:
============
These need to be run whenever the toolchain is modified or the latest nightly is needed.
* Update rust toolchain: `rustup update`
* Update dependencies: `cargo update`

Build dependencies, run the app and tests.
* Build dependencies: `cargo build`
* Run application `cargo run`
* Run all tests: `cargo test`

Database Migration:
-------------------
If you are adding database tables, a migration needs to be added. These are created and generated with the Rust ORM *Diesel*.

* Install sqlite-dev on your system `sudo dnf install sqlite-devel`
* Install diesel CLI: `cargo install diesel_cli --no-default-features --features sqlite`
* Setup diesel with database in repo (stored in .env): `diesel setup`
* Create migration files for new tables: `diesel migration generate <new_feature>`
* Modify the migration files and schema
* Test the migration applies with `diesel migration run` and that it can be rolled back with `diesel migration redo`

Troubleshooting:
----------------

* Run with backtrace enabled: `RUST_BACKTRACE=1 cargo <command>`
* Increase logging: `cargo run -- -v` (Repeat for each log level)
  * Log levels: error, warning, info, debug, trace


CHANGELOG.md
<a name="0.3.0"></a>
## 0.3.0 (2022-10-28)

Features
* Add PoolID and description to Pools page
* Improve logging during validation
* Add support for managing pools
* Update README to be a little more helpful for Rust newbies
* Username validation against Slack and Custom Owners support
* Make compatible with Rust 2018 edition
* Added pools endpoint
* Change device pool filtering to query string
* Add drop down for filtering devices by pool
* Add device table searching and hover highlighting
* Make reservations randomly choose a device
* Add pools to devices in the UI
* Add a reservations api endpoint for reserving devices
* Add pools for devices to be in
* Add API for ending device reservations

Fixes
* Ensure updating a device leaves users on the same page
* Allow edit of pool when non-empty

<a name="0.2.3"></a>
## 0.2.3 (2018-08-02)


#### Features

*   Add better validation error messages ([28d28807](https://github.com/tismith/device-checkout-rs/commit/28d288072a14711143d99d972a6ae100c1ceb4ea))



<a name="0.2.2"></a>
## 0.2.2 (2018-07-27)


#### Features

*   Add Docker packaging support ([cfb84c8f](https://github.com/tismith/device-checkout-rs/commit/cfb84c8fdad8ca2a092a22536d8f1a360343ff3e))
*   Add more context on errors ([6b9be7a1](https://github.com/tismith/device-checkout-rs/commit/6b9be7a16f66dcbd720655edd358bd5f8d805712))



<a name="0.2.1"></a>
## 0.2.1 (2018-06-06)


#### Bug Fixes

*   Set the pool size to 1, avoid SQLITE_BUSY errors ([ebcd3287](https://github.com/tismith/device-checkout-rs/commit/ebcd3287ba13200c6015e39e22ab9e7c79ed7841))

#### Features

*   Add a snap package manifest for distribution ([58aa3b7e](https://github.com/tismith/device-checkout-rs/commit/58aa3b7e418c42d9e2e65ef751bbb19c9bc70103))
*   Add a --templates command-line argument ([abfb73a3](https://github.com/tismith/device-checkout-rs/commit/abfb73a32a7258f733542aef90e115ad2a38ff66))



<a name="0.2.0"></a>
## 0.2.0 (2018-05-22)


#### Features

*   use the full display width for tables ([e8c57c54](https://github.com/tismith/device-checkout-rs/commit/e8c57c54c7bb13ed540f782c955f9df77b16e7de))
*   add confirmation dialogs to deletes ([92264047](https://github.com/tismith/device-checkout-rs/commit/922640477e6367a3f5e7c3cae7c9fd339de96cea))
*   use cookies to pass status messages around ([2961372b](https://github.com/tismith/device-checkout-rs/commit/2961372b54b503d38060ceeb0e0eb4fea9f3556d))



<a name="0.1.0"></a>
## 0.1.0 Device Checkout (2018-05-15)






src/utils/cmdline.rs
use clap;
use crate::utils::types;

pub fn parse_cmdline() -> types::Settings {
    let matches = matcher().get_matches();
    match parse(&matches) {
        Ok(s) => s,
        Err(e) => e.exit(),
    }
}

fn parse(matches: &clap::ArgMatches<'_>) -> Result<types::Settings, clap::Error> {
    let verbosity = matches.occurrences_of("verbosity") as usize;
    if verbosity > 4 {
        Err(clap::Error {
            message: "invalid number of 'v' flags".into(),
            kind: clap::ErrorKind::InvalidValue,
            info: None,
        })?
    }
    let quiet = matches.is_present("quiet");
    let timestamp = match matches.value_of("timestamp") {
        Some("ns") => types::Timestamp::Nanosecond,
        Some("ms") => types::Timestamp::Microsecond,
        Some("sec") => types::Timestamp::Second,
        Some("none") | None => types::Timestamp::Off,
        Some(_) => Err(clap::Error {
            message: "invalid value for 'timestamp'".into(),
            kind: clap::ErrorKind::InvalidValue,
            info: None,
        })?,
    };

    let template_dir = matches.value_of("templates").map(|x| x.to_string());

    let port = value_t!(matches.value_of("port"), u16)?;
    let database = matches.value_of("database").ok_or_else(|| clap::Error {
        message: "invalid value for 'database'".into(),
        kind: clap::ErrorKind::InvalidValue,
        info: None,
    })?;

    Ok(types::Settings {
        verbosity,
        quiet,
        timestamp,
        port,
        template_dir,
        database_url: database.to_string(),
        ..Default::default()
    })
}

fn matcher<'a, 'b>() -> clap::App<'a, 'b> {
    clap::App::new(crate_name!())
        .version(crate_version!())
        .author(crate_authors!())
        .arg(
            clap::Arg::with_name("verbosity")
                .short("v")
                .multiple(true)
                .help("Increase message verbosity, maximum 4"),
        )
        .arg(
            clap::Arg::with_name("quiet")
                .short("q")
                .long("quiet")
                .help("Silence all output"),
        )
        .arg(
            clap::Arg::with_name("timestamp")
                .short("t")
                .long("timestamp")
                .help("prepend log lines with a timestamp")
                .takes_value(true)
                .possible_values(&["none", "sec", "ms", "ns"]),
        )
        .arg(
            clap::Arg::with_name("port")
                .short("p")
                .long("port")
                .help("tcp port number to listen on")
                .default_value("8000")
                .takes_value(true),
        )
        .arg(
            clap::Arg::with_name("templates")
                .long("templates")
                .help("directory for templates")
                .takes_value(true),
        )
        .arg(
            clap::Arg::with_name("database")
                .short("d")
                .long("database")
                .help("path to database file")
                .default_value("devices.db")
                .takes_value(true),
        )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_port() {
        let m = matcher()
            .get_matches_from_safe(vec!["", "--port", "1234"])
            .unwrap();
        let s = parse(&m).unwrap();

        assert_eq!(s.port, 1234u16);
    }

    #[test]
    fn test_database() {
        let m = matcher()
            .get_matches_from_safe(vec!["", "--database", "somefile.txt"])
            .unwrap();
        let s = parse(&m).unwrap();

        assert_eq!(s.database_url, "somefile.txt".to_string());
    }

    #[test]
    fn test_template_dir() {
        let m = matcher()
            .get_matches_from_safe(vec!["", "--templates", "somedir"])
            .unwrap();
        let s = parse(&m).unwrap();

        assert_eq!(s.template_dir, Some("somedir".to_string()));
    }

    #[test]
    fn test_no_template_dir() {
        let m = matcher().get_matches_from_safe(vec![""]).unwrap();
        let s = parse(&m).unwrap();

        assert_eq!(s.template_dir, None);
    }

    #[test]
    fn test_too_much_verbosity() {
        let m = matcher().get_matches_from_safe(vec!["", "-vvvvv"]).unwrap();
        assert!(parse(&m).is_err());
    }

    #[test]
    fn test_just_enough_verbosity() {
        let m = matcher().get_matches_from_safe(vec!["", "-vvv"]).unwrap();
        let s = parse(&m).unwrap();

        assert_eq!(s.verbosity, 3);
    }

    #[test]
    fn test_timestamps() {
        let m = matcher()
            .get_matches_from_safe(vec!["", "-t", "sec"])
            .unwrap();
        let s = parse(&m).unwrap();

        match s.timestamp {
            types::Timestamp::Second => (),
            _ => panic!("unexpected parse"),
        }
    }

    #[test]
    fn test_bogus_timestamps() {
        assert!(matcher()
            .get_matches_from_safe(vec!["", "-t", "bogus"])
            .is_err());
    }
}
